quickjs.c

```diff
@@ -50132,3 +50417,168 @@ void JS_AddPerformance(JSContext *ctx)
+    InlineCacheHashSlot *ch, *ch_next;
+    InlineCacheHashSlot **new_hash;
+    ic->hash_bits += 1;
+    new_capacity = 1 << ic->hash_bits;
+    new_hash = js_malloc(ic->ctx, sizeof(ic->hash[0]) * new_capacity);
+    if (unlikely(!new_hash))
+        goto fail;
+    memset(new_hash, 0, sizeof(ic->hash[0]) * new_capacity);
+    for (i = 0; i < ic->capacity; i++) {
+        for (ch = ic->hash[i]; ch != NULL; ch = ch_next) {
+            h = get_index_hash(ch->atom, ic->hash_bits);
+            ch_next = ch->next;
+            ch->next = new_hash[h];
+            new_hash[h] = ch;
+        }
+    }
+    js_free(ic->ctx, ic->hash);
+    ic->hash = new_hash;
+    ic->capacity = new_capacity;
+    return 0;
+fail:
+    return -1;
+}
+
+int free_ic(InlineCache *ic)
+{
+    uint32_t i, j;
+    InlineCacheHashSlot *ch, *ch_next;
+    InlineCacheRingItem *buffer;
+    if (ic->cache) {
+        for (i = 0; i < ic->count; i++) {
+            buffer = ic->cache[i].buffer;
+            JS_FreeAtom(ic->ctx, ic->cache[i].atom);
+            for (j = 0; j < IC_CACHE_ITEM_CAPACITY; j++) {
+                js_free_shape_null(ic->ctx->rt, buffer[j].shape);
+            }
+        }
+    }
+    for (i = 0; i < ic->capacity; i++) {
+        for (ch = ic->hash[i]; ch != NULL; ch = ch_next) {
+            ch_next = ch->next;
+            JS_FreeAtom(ic->ctx, ch->atom);
+            js_free(ic->ctx, ch);
+        }
+    }
+    if (ic->count > 0)
+        js_free(ic->ctx, ic->cache);
+    js_free(ic->ctx, ic->hash);
+    js_free(ic->ctx, ic);
+    return 0;
+}
+
+uint32_t add_ic_slot(InlineCache *ic, JSAtom atom, JSObject *object,
+                     uint32_t prop_offset)
+{
+    int32_t i;
+    uint32_t h;
+    InlineCacheHashSlot *ch;
+    InlineCacheRingSlot *cr;
+    JSShape *sh;
+    cr = NULL;
+    h = get_index_hash(atom, ic->hash_bits);
+    for (ch = ic->hash[h]; ch != NULL; ch = ch->next)
+        if (ch->atom == atom) {
+            cr = ic->cache + ch->index;
+            break;
+        }
+
+    assert(cr != NULL);
+    i = cr->index;
+    for (;;) {
+        if (object->shape == cr->buffer[i].shape) {
+            cr->buffer[i].prop_offset = prop_offset;
+            goto end;
+        }
+        i = (i + 1) % IC_CACHE_ITEM_CAPACITY;
+        if (unlikely(i == cr->index))
+            break;
+    }
+    sh = cr->buffer[i].shape;
+    cr->buffer[i].shape = js_dup_shape(object->shape);
+    js_free_shape_null(ic->ctx->rt, sh);
+    cr->buffer[i].prop_offset = prop_offset;
+end:
+    return ch->index;
+}
+
+uint32_t add_ic_slot1(InlineCache *ic, JSAtom atom)
+{
+    uint32_t h;
+    InlineCacheHashSlot *ch;
+    if (ic->count + 1 >= ic->capacity && resize_ic_hash(ic))
+        goto end;
+    h = get_index_hash(atom, ic->hash_bits);
+    for (ch = ic->hash[h]; ch != NULL; ch = ch->next)
+        if (ch->atom == atom)
+            goto end;
+    ch = js_malloc(ic->ctx, sizeof(InlineCacheHashSlot));
+    if (unlikely(!ch))
+        goto end;
```

Comment:
For my understanding, it's safe and sound to bail out because it only results in a cache miss, no other observable changes?

Reply:
Yes